lib/Sema/CSSimplify.cpp

```diff
@@ -2811,6 +2811,20 @@ ConstraintSystem::matchTypes(Type type1, Type type2, ConstraintKind kind,
                                          formUnsolvedResult);
         }
       }
+      
+      // If we are trying to perform coercion to the same type emit a warning.
+      if (locator.getBaseLocator()->isTypeCoercion()) {
```

Comment:
@xedin is right: CSSimplify is too early to be diagnosing this.  The new diagnostics infrastructure will help to correctly defer this out to the right place.

Reply:
@LucianoPAlmeida I think you'd want to do this somewhere like https://github.com/apple/swift/pull/26710/files#diff-69306f7b0822cf52e8c0639307ad7407R3092. As you can see a bit later on this handles a cases where either one of the types is a type variable or both of them are, which means that actual comparison has to be differed until both types are bound.

Reply:
Also `locator.getBaseLocator()->isTypeCoercion()` is not the right check, you'd want to check whether locator ends at `isTypeCoercion` regardless of it having any implicit conversions in-between or originating from locator. I'd suggest to wait a bit until https://github.com/apple/swift/pull/26297 lands, which would add `findFirst` and make this much easier.

Reply:
Hey @xedin @CodaFi, just trying to better understand this: so we could do this check here if there is a fixed type for type1 and type2, but is one of them e.g. type1 is a typevar we should deffer that to later on the pipeline .e.g matchTypesBindTypeVar? 
Also when handling ConstraintKind::Conversion here, is there any case where it has a fixed type at this point in CSSimplify?

Reply:
Originally recorded binary constraints have a type variable on either side e.g. `$T0 conv Int` or `Int argument conv $T0`, otherwise they can be simplified upon creation, as solver runs it would try to infer potential type(s) for each type variable, bind it, and try to simplify associated constraints.

So when it sees e.g. `argument conversion` and one side is still a type variable it just means that it can't simplify such constraint until that type variable is bound to some type. For this fix we need to make sure that both sides of the conversion constraint are bound.

To better understand what is going on in the solver you could try to run compiler with `-debug-constraints` flag which would output information about everything solver does.

Reply:
Thank you so much for all the info @xedin :)) This is still in WIP, but I expect to get something ready soon 👍